informers: Don't treat relist same as sync

squeed opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

### Background:

Before this change, DeltaFIFO emits the Sync DeltaType on Resync() and Replace(). Seperately, the SharedInformer will only pass that event on to handlers that have a ResyncInterval and are due for Resync. This can cause updates to be lost if an object changes as part of the Replace(), as it may be incorrectly discarded if the handler does not want a Resync.

### What this change does:

Creates a new DeltaType, Replaced, which is emitted by DeltaFIFO on Replace(). For backwards compatability concerns, the old behavior of always emitting Sync is preserved unless explicity overridden.

As a result, if an object changes (or is added) on Replace(), now all SharedInformer handlers will get a correct Add() or Update() notification.

One additional side-effect is that handlers which do not ever want Resyncs will now see them for all objects that have not changed during the Replace.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Shared informers are now more reliable in the face of network disruption.
```

/cc @sttts 
/cc @liggitt

squeed committed:
client-go/cache/testing: add ability to simulate watch disruption

This adds ResetWatch() to the FakeControllerSource, which lets the
controller simulate a re-list-and-watch.

squeed commented:
Many thanks to @sttts, who helped me understand exactly where to look in the SharedInformer when faced with an "impossible" sequence of log lines.

liggitt commented:
> Separately, there is a commit that elides re-List updates if RV and UID hasn't changed.

Can that be split to a separate PR? That's separate from the bug fix, correct? We've previously decided against teaching informers about uids, would like to consider that separately and not backport that.

abhat commented on staging/src/k8s.io/client-go/tools/cache/testing/fake_controller_source.go:
```diff
@@ -75,6 +78,16 @@ type nnu struct {
 	uid             types.UID
 }
 
+// SwapWatch simulates connection problems; creates a new Broadcaster and flushes
```
nitpick: `s/SwapWatch/ResetWatch`

squeed replied:
fixed

abhat commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -456,6 +457,10 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 			isSync := d.Type == Sync
 			s.cacheMutationDetector.AddObject(d.Object)
 			if old, exists, err := s.indexer.Get(d.Object); err == nil && exists {
+				// If the delta is from a relist, and the RV hasn't changed, elide it.
```
We also compare UIDs. Comment can be updated to reflect the behavior of `isObjectSame`

squeed replied:
removed the UID comparison.

squeed commented:
> Can that be split to a separate PR? That's separate from the bug fix, correct? We've previously decided against teaching informers about uids, would like to consider that separately and not backport that.

It's not separate to this bug fix; @sttts suggested that I split it out since it's not necessary. There are two points for consideration:

1. The larger bugfix is also changing existing behavior. Assume a handler without ResyncInterval doesn't want to get a Sync-style level (e.g. Update(A1, A1) and only wants edges. However, if the client _internally_ decides to do a re-List, then the handler will still see no-op updates. Now, I don't think there's any kind of contract that says you'll *never* get no-op updates, but it it was coded in to some of the tests.

2. As to UIDs, I wasn't sure of the guarantees for ResourceVersion across delete-and-recreate. It's not explicitly mentioned in the documentation. If a new object can't have the same ResourceVersion (currently impossible with etcd, of course), then we can drop the UID.

liggitt commented:
> 2\. If a new object can't have the same ResourceVersion (currently impossible with etcd, of course), then we can drop the UID.

resourceVersion is unique per resource type (it has to be, since we use it for list requests across resource instances), so it cannot repeat on a delete/recreate

liggitt commented:
> It's not separate to this bug fix; @sttts suggested that I split it out since it's not strictly necessary.

We should split it out, as it needs more discussion, and even if accepted, I would not expect to backport that part.

squeed commented:
> resourceVersion is unique per resource type (it has to be, since we use it for list requests across resource instances), so it cannot repeat on a delete/recreate

Ack. Removed UID check; doesn't affect correctness.

liggitt commented:
fyi @smarterclayton @lavalamp

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -357,7 +357,7 @@ func newInformer(
 			for _, d := range obj.(Deltas) {
 				switch d.Type {
-				case Sync, Added, Updated:
+				case Sync, Replaced, Added, Updated:
```
do we support external consumers of controller.Config / Process? this PR updates our two internal implementations, but would break all external uses, correct?

squeed replied:
I don't see anything that references these types (specifically cache.Config or cache.Deltas) outside the package. They probably should have been private types...

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -577,10 +577,11 @@ const (
-	// (Anything that trigger's DeltaFIFO's Replace() method.)
+	// Replaced is emitted when we encountered watch errors and had to do a
+	// relist. We don't here know if the newly relisted object has changed;
+	// shared informers will compare against their store to find out.
```
it's not great to describe shared informer behavior here, this layer isn't supposed to know about them

squeed replied:
@sttts told me to add this - but I agree it's a layer violation :-)

squeed commented:
> We should split it to a separate PR, as it needs more discussion, and even if accepted, I would not expect to backport that part.

Understood. The open question to me is: is it OK to send resync updates to handlers that don't want them? If so, then we don't need the last commit. Otherwise we do.

lavalamp commented:
I think it is probably good to make the distinction, but this will be disruptive to users, potentially very disruptive.

Therefore, I think you need to wire an option or provide a 2nd constructor or something, so that existing code can continue to work the same with no changes.

squeed commented:
> I think it is probably good to make the distinction, but this will be disruptive to users, potentially very disruptive.
> 
> Therefore, I think you need to wire an option or provide a 2nd constructor or something, so that existing code can continue to work the same with no changes.

@lavalamp I don't quite follow - what sort of option were you thinking of?

To be clear here, this PR doesn't disable resyncs or change the standard informer behavior at all. It just ensures we don't miss updates due to a re-List. AFAICT, the question is whether or not we should expose the re-List due to disruption to the handlers.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -577,10 +577,11 @@ const (
+	// Replaced is emitted when we encountered watch errors and had to do a
+	// relist. We don't here know if the newly relisted object has changed;
+	// shared informers will compare against their store to find out.
+	Replaced DeltaType = "Replaced"
```
Old client may have written code that will not expect to see a new DeltaType.

squeed replied:
Ah. Yikes.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -445,7 +445,7 @@ func (f *DeltaFIFO) Replace(list []interface{}, resourceVersion string) error {
 		}
 		keys.Insert(key)
-		if err := f.queueActionLocked(Sync, item); err != nil {
+		if err := f.queueActionLocked(Replaced, item); err != nil {
```
```
action := Sync
if f.clientUnderstandsReplacedDeltaType {
  action = Replaced
}
if err := f.queueActionLocked(action, item); err != nil {
```

squeed replied:
*nod*

liggitt replied:
agree, and can set that to true in both places we call NewDeltaFIFO in-tree (in newInformer and sharedIndexInformer#Run)

lavalamp commented:
> doesn't disable resyncs or change the standard informer behavior at all

There can be direct clients of this.

The fact that we can make corresponding changes in the informer to preserve its behavior makes your life much easier though.

squeed commented:
As an alternative to adding a new Delta type, we could "fix" this in the SharedInformer by cheating a bit, and treating Sync as Updated if the object differs from the store.

squeed commented:
hey @lavalamp thanks for the commentary, I've filed a PR with an alternate approach in #86114

fedebongio commented:
/assign @deads2k @liggitt

squeed commented:
Incidentally, we may be able to put more cases in `dedupDeltas()` to avoid unnecessary updates. Though, of course, potentially missing updates is much worse.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -452,10 +453,14 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+				// If the delta is from a relist, and the RV hasn't changed, elide it.
+				if d.Type == Replaced && isObjectSame(old, d.Object) {
+					continue
+				}
```
is this fundamental to the fix, or a separate optimization? I expected just the change in how we compute `isSync`

squeed replied:
I would argue it's important. If a handler has requested no resyncs (i.e. no levels, just edges) we should presumably respect that.

squeed replied:
But, yes, we could set to isSync rather than eliding.

liggitt replied:
> If a handler has requested no resyncs (i.e. no levels, just edges) we should presumably respect that.

That's not how I would have explained resyncs... I always took the expressed resync period to mean an artificial replay of all existing items in the store as update events, *not* the events you got on a relist which was required because the watch dropped and couldn't be restarted.

> we could alter isSync rather than eliding entirely

The PR has already altered isSync by introducing a new delta type... I thought that was sufficient to fix the observed bug

liggitt replied:
Hmm, I see `processorListener#shouldResync` always returned false if a particular listener didn't request a resync, so those listeners would not have previously gotten any events from a relist (which got propagated as Sync delta events).

So, this currently changes the following behaviors:
* listeners that did not request a resync period now get notified of updated objects on relist (good)
* listeners that did request a resync period no longer get notified of unchanged objects on relist (questionable)

squeed replied:
> I see processorListener#shouldResync always returned false

Even worse, `sharedProcessor#shouldResync` is side-effecting! While not an issue per-se, side-effecting predicates are ... unexpected ...

> listeners that did request a resync period no longer get notified of unchanged objects on relist (questionable)

IMO, re-List / Replace is an interaction local to the reflector and its store (DeltaFIFO in this case). It should be DeltaFIFO's responsibility to generate the correct sequence of Added / Updated / Deleted / Sync deltas based on what it gets from the reflector. 

It doesn't seem valuable that SharedInformer handlers get a resync (of all objects) sooner than their ResyncInterval would suggest if the reflector does a re-List. It smells like a layer violation.

squeed committed:
informers: don't treat relist same as sync

Background:

Before this change, DeltaFIFO emits the Sync DeltaType on Resync() and
Replace(). Seperately, the SharedInformer will only pass that event
on to handlers that have a ResyncInterval and are due for Resync. This
can cause updates to be lost if an object changes as part of the Replace(),
as it may be incorrectly discarded if the handler does not want a Resync.

What this change does:

Creates a new DeltaType, Replaced, which is emitted by DeltaFIFO on
Replace(). For backwards compatability concerns, the old behavior of
always emitting Sync is preserved unless explicity overridden.

As a result, if an object changes (or is added) on Replace(), now all
SharedInformer handlers will get a correct Add() or Update()
notification.

One additional side-effect is that handlers which do not ever want
Resyncs will now see them for all objects that have not changed during
the Replace.

squeed commented:
Thanks all for the feedback. I've updated the PR based on review comments.  There is now a `DeltaFIFO.clientUnderstandsReplaced` and the corresponding wiring.

@lavalamp @liggitt PTAL.

squeed commented:
(verify failure is a misspelling; I'll fix that with the next round of review).

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -344,7 +344,10 @@ func newInformer(
 	// This will hold incoming changes. Note how we pass clientState in as a
 	// KeyLister, that way resync operations will result in the correct set
 	// of update/delete deltas.
-	fifo := NewDeltaFIFO(MetaNamespaceKeyFunc, clientState)
```
why no longer provide the keyFunc?

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -57,14 +57,53 @@ import (
+	// the Replaced DeltaType. Before Replaced was added, Replace() was treated
+	// the same as Sync(). For backwards-compatability purposes, this is false
+	// by default.
+	ClientUnderstandsReplaced bool
```
This approach lgtm.

nit: I would be more explicit about the name (e.g. `ClientUnderstandsReplacedDeltaType`)

squeed commented:
I'd like to clean up `delta_fifo_test.go` to use the same `NewDeltaFIFO` as the `SharedInformer` in a followup commit, once things generally look good.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -57,14 +57,53 @@ import (
 //
 // Also see the comment on DeltaFIFO.
```
since callers of this method are potentially at risk, we should mark it deprecated in godoc and document the recommended invocation of NewDeltaFIFOWithOptions:

```
// Warning: This constructs a DeltaFIFO that does not differentiate between events caused by a call to Replace
// (for example, from a relist, which may contain object updates), and synthetic events caused by a periodic resync
// (which just emit the existing object). See https://issue.k8s.io/86015 for details.
//
// Use `NewDeltaFIFOWithOptions(DeltaFIFOOptions{..., ClientUnderstandsReplacedDeltaType: true})` instead
// to receive a `Replaced` event depending on the type.
//
// Deprecated: Equivalent to NewDeltaFIFOWithOptions(DeltaFIFOOptions{KeyFunction: keyFunc, KnownObjects: knownObjects});
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:
```diff
@@ -98,14 +98,15 @@ func TestDeltaFIFO_replaceWithDeleteDeltaIn(t *testing.T) {
 	f := NewDeltaFIFO(testFifoObjectKeyFunc, keyLookupFunc(func() []testFifoObject {
 		return []testFifoObject{oldObj}
 	}))
+	f.clientUnderstandsReplaced = true
```
I would actually like to leave this test entirely as-is, and make a copy that sets this option and verifies the Replaced delta type

liggitt replied:
same comment applies to the other existing unit tests that were modified

liggitt commented on staging/src/k8s.io/client-go/tools/cache/testing/fake_controller_source_test.go:
```diff
@@ -93,3 +93,37 @@ func TestRCNumber(t *testing.T) {
 	wg.Wait()
 }
+
+func TestRCNumberDisruptive(t *testing.T) {
```
comments explaining what this is testing would be helpful to future generations... it's not immediately clear to me how this relates to events from a relist

squeed commented:
@liggitt updated based on review feedback. I'll be somewhat less reachable over the coming days (parental leave), but I'll try and respond to this PR.

tedyu commented on staging/src/k8s.io/client-go/tools/cache/testing/fake_controller_source.go:
```diff
@@ -229,9 +239,27 @@ func (f *FakeControllerSource) Watch(options metav1.ListOptions) (watch.Interfac
+		// get the RV of the oldest object in the change queue
+		oldestRV, err := meta.NewAccessor().ResourceVersion(f.changes[0].Object)
+		if err != nil {
+			panic(err)
```
Should err be returned instead of panic'ing ?

liggitt replied:
No, errors have the potential to trigger different flows in the client handling. This indicates a programming error in the test, and we just want to stop the test at this point.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+type DeltaFIFOOptions struct {
+
+	// KeyFunction is used to figure out what key an object should have. (It's
+	// exposed in the returned DeltaFIFO's KeyOf() method, with bonus features.)
```
> with bonus features

that's ominous :)
should probably explain what is different

squeed replied:
Don't blame me, I just copied the doctext from NewDeltaFIFO. I'll add a quick explanation.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+
+	// KnownObjects is expected to return a list of keys that the consumer of
+	// this queue "knows about". It is used to decide which items are missing
+	// when Replace() is called; 'Deleted' deltas are produced for these items.
```
```suggestion
	// when Replace() is called; 'Deleted' deltas are produced for the missing items.
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+	// KnownObjects is expected to return a list of keys that the consumer of
+	// this queue "knows about". It is used to decide which items are missing
+	// when Replace() is called; 'Deleted' deltas are produced for these items.
+	// It may be nil if you don't need to detect all deletions.
```
```suggestion
	// KnownObjects may be nil if you don't need to detect deletions.
```

squeed replied:
If you don't set KnownObjects, you *do* get deletions, except on Replace. So it's more subtle.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+	// It may be nil if you don't need to detect all deletions.
+	KnownObjects KeyListerGetter
+
+	// ClientUnderstandsReplacedDeltaType is whether or not the queue consumer understands
```
```suggestion
	// ClientUnderstandsReplacedDeltaType indicates the queue consumer understands
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+	KnownObjects KeyListerGetter
+
+	// ClientUnderstandsReplacedDeltaType is whether or not the queue consumer understands
+	// the Replaced DeltaType. Before Replaced was added, Replace() was treated
```
```suggestion
	// the Replaced DeltaType. Before the `Replaced` event type was added, calls to Replace() were handled
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+	// ClientUnderstandsReplacedDeltaType is whether or not the queue consumer understands
+	// the Replaced DeltaType. Before Replaced was added, Replace() was treated
+	// the same as Sync(). For backwards-compatibility purposes, this is false
+	// by default.
```
```suggestion
	// by default. When true, `Replaced` events for items passed to a Replace() call.
	// When false, `Sync` events will be sent for items passed to a Replace() call.
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,15 +56,64 @@ import (
+	// KeyFunction is used to figure out what key an object should have. (It's
+	// exposed in the returned DeltaFIFO's KeyOf() method, with bonus features.)
+	// Optional, the default is MetaNamespaceKeyFunc which is essentially
+	// <namespace>/<name>
```
> `which is essentially <namespace>/<name>`

crisp this up, either describe exactly what it is (I assume the details are around handling of non-namespaced objects?), or just say we default to `MetaNamespaceKeyFunc`

liggitt commented:
Code looks good. Some godoc nits, then lgtm.

squeed commented:
@liggitt Updated and rebased, thanks for the feedback.

liggitt commented:
/retest

liggitt commented:
/lgtm
/approve

liggitt commented:
unit test failure is legitimate:
```
$ go test ./vendor/k8s.io/client-go/tools/cache
# k8s.io/client-go/tools/cache [k8s.io/client-go/tools/cache.test]
staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:294:3: undefined: keyLookupFunc
staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:410:17: undefined: keyLookupFunc
staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:440:17: undefined: keyLookupFunc
```

liggitt commented:
/lgtm cancel

squeed commented:
whoops, fixing

squeed commented:
/retest

liggitt commented:
/retest

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -56,12 +57,62 @@ import (
+	// backwards-compatibility purposes, this is false by default.
+	// When true, `Replaced` events will be sent for items passed to a Replace() call.
+	// When false, `Sync` events will be sent instead.
+	ClientUnderstandsReplacedDeltaType bool
```
nit: I'd go with a different name, an uncareful reader might think there's some server logic being triggered. Ideas:

* PermitDeltaTypeReplaced
* SeparateSyncAndReplace

squeed replied:
Ah yeah, the notion of "client" here is not really correct. Hows about "EmitDeltaTypeReplaced" ?

lavalamp replied:
That's probably fine, it solves the problem I mentioned.

liggitt replied:
works for me

lavalamp commented:
Overall, great change-- I hate to ask for something but the name is unchangeable once we cut a client-go release, so...

squeed commented:
OK, rename done.

liggitt commented:
/lgtm
/approve
/retest
